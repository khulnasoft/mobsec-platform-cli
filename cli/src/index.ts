/*
 * Copyright Â© 2023 KhulnaSoft Ltd.
 *
 * SPDX-License-Identifier: MIT
 */

export { run } from "@oclif/core";
