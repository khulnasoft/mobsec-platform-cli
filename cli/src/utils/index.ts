/*
 * Copyright Â© 2023 KhulnaSoft Ltd.
 *
 * SPDX-License-Identifier: MIT
 */

export * from "./args";
export * from "./app-ref";
export * from "./base-command";
export * from "./command-core";
export * from "./config";
export * from "./filters";
export * from "./group-list";
export * from "./parsers";
export * from "./sub-command-lister";
export * from "./table";
