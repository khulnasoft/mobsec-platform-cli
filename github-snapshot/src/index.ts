/*
 * Copyright Â© 2022 MobSec Inc.
 *
 * SPDX-License-Identifier: MIT
 */

import type { DependencySnapshot, Manifest } from "./dependency-snapshot";
import type { DependenciesOutput } from "@mobsec/platform-lib/lib/types";

/** The default GitHub REST server */
export const GITHUB_API_URL = "https://api.github.com";

/**
 * Upload URL for snapshots.
 *
 * @param owner Repository owner
 * @param repo Repository name
 * @param apiUrl optional: The address of the GitHub API server
 * @returns the url to use in an HTTP POST to upload the snapshot
 */
export const snapshotUploadUrl = (
  owner: string,
  repo: string,
  apiUrl = GITHUB_API_URL
) => `${apiUrl}/repos/${owner}/${repo}/dependency-graph/snapshots`;

/**
 * Generate a package url (PURL)
 */
function encodePurl(ecosystem: string, name: string, version?: string): string {
  let purl = `pkg:${ecosystem}/${name}`;
  if (version) {
    purl += `@${version}`;
  }

  return purl;
}

/**
 * Convert a Platform report to the Snapshot format.
 *
 * ref, sha and runId can be retrieved from the action context
 * ```
 * import * as github from "@actions/github";
 * const {ref, sha, runId} = github.context
 * ```
 *
 * @param dependencies The dependency data retrieved from the assessment.
 * `fetchAssessment` or `pollForReport` in `@mobsec/platform-lib` retrieves
 * this data in the correct format
 * @param githubCorrelator Correlation id to group snapshots over time.
 * @param sha commit hash of the commit that triggered the snapshot
 * @param ref Ref of the git commit. For example, "refs/heads/main".
 * @param runId runId of the action
 * @returns a dependency snapshot that can be uploaded to GitHub
 */
export function convertToSnapshot(
  dependencies: DependenciesOutput,
  githubCorrelator: string,
  sha: string,
  ref: string,
  runId: number
): DependencySnapshot {
  const manifests = new Map<string, Manifest>();

  for (const component of dependencies.components) {
    let source = component.source;
    if (source.startsWith("/")) {
      source = source.substring(1);
    }

    let manifest = manifests.get(source);
    if (!manifest) {
      manifest = {
        name: source,
        resolved: {},
      };
    }

    // We need to translate our ecosystem into a proper PURL type.
    // See https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst for supported PURL
    // types.
    let ecosystem = component.ecosystem.toLowerCase();
    switch (component.ecosystem) {
      case "Commercial":
      case "Native":
        // Fallback to "generic" which is reserved for this purpose.
        ecosystem = "generic";
        break;
      case "CocoaPods":
      case "Maven":
      case "npm":
      case "NuGet":
        break;
      // We purposefully omit default because the typings include a union here so if new types are
      // added the TypeScript compiler will emit an error.
    }

    const purl = encodePurl(ecosystem, component.name, component.version);

    manifest.resolved[component.name] = {
      package_url: purl.toString(),
      relationship: "direct",
    };

    manifests.set(source, manifest);
  }

  const snapshot: DependencySnapshot = {
    version: 0,
    ref,
    sha,
    job: {
      correlator: githubCorrelator,
      id: `${runId}`,
    },
    detector: {
      name: "MobSec",
      url: "https://www.khulnasoft.com/",
      version: dependencies.deputy,
    },
    scanned: new Date(dependencies.timestamp).toISOString(),
    manifests: Object.fromEntries(manifests),
  };

  return snapshot;
}
