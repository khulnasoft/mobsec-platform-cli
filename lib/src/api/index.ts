/*
 * Copyright Â© 2022 MobSec Inc.
 *
 * SPDX-License-Identifier: MIT
 */

export * from "./app";
export * from "./assessment";
export * from "./organization";
export * from "./user";
