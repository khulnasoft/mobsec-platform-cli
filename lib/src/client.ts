/*
 * Copyright Â© 2022 MobSec Inc.
 *
 * SPDX-License-Identifier: MIT
 */

import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import axiosRetry from "axios-retry";

import { DocumentNode } from "graphql/language/ast";
import { GraphQLErrorData, GraphQLResponse } from "./types/graphql";
import { version } from "./version";
import {
  PlatformConfig,
  DEFAULT_API_URL,
  DEFAULT_LAB_API_URL,
} from "./platform-config";
import { CustomError } from "./errors";

export type RequestConfig = Pick<
  AxiosRequestConfig,
  | "onDownloadProgress"
  | "onUploadProgress"
  | "signal"
  | "timeout"
  | "timeoutErrorMessage"
>;

export type JSONType =
  | string
  | number
  | boolean
  | null
  | { [index: string]: JSONType }
  | JSONType[];
export type JSONObject = { [index: string]: JSONType };
export type JSONArray = JSONType[];

export const USER_AGENT = `MobSec API library ${version}`;

export class GraphQLError extends CustomError {
  constructor(
    public readonly errors: GraphQLErrorData[],
    public readonly data: JSONObject
  ) {
    super(`graphql error: ${errors.map((err) => err.message).join(", ")}`);
  }
}

export function getGqlString(doc: DocumentNode) {
  return doc.loc && doc.loc.source.body;
}

export interface NSClientOptions {
  userAgent?: string;
}
export class MobSecClient {
  private client: AxiosInstance;
  private apiUrl: string;
  private labApiUrl: string;
  constructor(platform: PlatformConfig, clientOptions?: NSClientOptions) {
    const {
      token: platformToken,
      apiUrl = DEFAULT_API_URL,
      labApiUrl = DEFAULT_LAB_API_URL,
    } = platform;
    this.apiUrl = apiUrl;
    this.labApiUrl = labApiUrl;

    // custom instance to keep our setup from polluting the main axios setup
    this.client = axios.create({
      headers: {
        Authorization: `Bearer ${platformToken}`,
        "User-Agent": clientOptions?.userAgent || USER_AGENT,
      },
    });
    // Retry idempotent operations up to 3 times. Delay is set to zero for tests
    axiosRetry(this.client, {
      retries: 3,
      retryDelay: (retryCount) =>
        process.env.NODE_ENV !== "test" ? retryCount * 1000 + 100 : 0,
    });
  }

  async graphqlRequest<T extends JSONObject, V extends JSONObject = JSONObject>(
    query: string | DocumentNode,
    variables: V = null,
    ignoreGQLErrors = false,
    config?: RequestConfig,
    isMutation = false
  ): Promise<T> {
    if (typeof query !== "string") {
      query = getGqlString(query);
    }
    const request: AxiosRequestConfig = {
      ...(config || {}),
      baseURL: this.apiUrl,
      url: "/graphql",
    };

    if (isMutation) {
      request.method = "post";
      request.data = { query, variables: variables || {} };
    } else {
      request.method = "get";
      request.params = {
        query,
        variables: JSON.stringify(variables || {}),
      };
    }

    const response = await this.client<GraphQLResponse<T>>(request);

    if (response.data.errors && !ignoreGQLErrors) {
      throw new GraphQLError(response.data.errors, response.data.data);
    }

    return response.data.data || null;
  }

  async query<T extends JSONObject, V extends JSONObject = JSONObject>(
    query: string | DocumentNode,
    variables: V = null,
    ignoreGQLErrors = false,
    config?: RequestConfig
  ): Promise<T> {
    return this.graphqlRequest(
      query,
      variables,
      ignoreGQLErrors,
      config,
      false
    );
  }

  async mutation<T extends JSONObject, V extends JSONObject = JSONObject>(
    query: string | DocumentNode,
    variables: V = null,
    ignoreGQLErrors = false,
    config?: RequestConfig
  ): Promise<T> {
    return this.graphqlRequest(query, variables, ignoreGQLErrors, config, true);
  }

  async get<T>(relativeUrl: string, config?: RequestConfig) {
    const response = await this.client.get<T>(
      this.restURL(relativeUrl),
      config
    );

    return response.data;
  }

  async post<T>(
    relativeUrl: string,
    params: JSONObject = {},
    config?: RequestConfig
  ) {
    const response = await this.client.post<T>(
      this.restURL(relativeUrl),
      params,
      config
    );

    return response.data;
  }

  async postStream<T>(
    relativeUrl: string,
    stream: NodeJS.ReadableStream,
    config?: RequestConfig
  ): Promise<T> {
    // Fix for https://github.com/axios/axios/issues/1045 - if maxRedirects is anything
    // other than zero the entire stream will be read into memory.
    const fullConfig: AxiosRequestConfig = {
      ...(config || {}),
      maxRedirects: 0,
    };
    const response = await this.client.post<T>(
      this.restURL(relativeUrl),
      stream,
      fullConfig
    );

    return response.data;
  }

  async delete<T = void>(
    relativeUrl: string,
    config?: RequestConfig
  ): Promise<T> {
    const response = await this.client.delete(
      this.restURL(relativeUrl),
      config
    );
    return response.data;
  }

  restURL(relativeUrl: string) {
    return new URL(relativeUrl, this.labApiUrl).href;
  }

  // For testing purposes only
  _axiosClient() {
    return this.client;
  }
}
