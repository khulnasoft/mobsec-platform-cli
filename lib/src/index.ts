/*
 * Copyright Â© 2022 MobSec Inc.
 *
 * SPDX-License-Identifier: MIT
 */

export * from "./api";
export * from "./client";
export * from "./errors";
export * from "./findings-filter";
export * from "./finding-keys";
export * from "./platform";
export * from "./platform-config";
export * from "./utils";
export * from "./version";
