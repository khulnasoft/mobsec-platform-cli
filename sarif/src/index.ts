/*
 * Copyright Â© 2021-2022 MobSec Inc.
 *
 * SPDX-License-Identifier: MIT
 */

import type {
  Log,
  Run,
  ReportingDescriptor,
  Result,
  Notification,
  PhysicalLocation,
} from "sarif";
import { ripGrep as rg, RipGrepError } from "ripgrep-js";

import type { Assessment, Severity } from "@mobsec/platform-lib/lib/types";
import {
  Filter,
  findingMatchesFilter,
  DEFAULT_FILTER,
  KeyParams,
  DEFAULT_KEY_PARAMS,
  ruleIdAndFingerprint,
} from "@mobsec/platform-lib";
import {
  DEFAULT_LAB_UI_URL,
  assessmentLink,
} from "@mobsec/platform-lib/lib/platform-config";

export { version } from "./version";

const SARIF_SCHEMA_URL =
  "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json";

/**
 * Convert MobSec severity to a SARIF notification level.
 */
function severityToNotification(input: string): Notification.level {
  if (input === "high" || input === "critical") {
    return "error";
  } else if (input === "medium" || input === "low") {
    return "warning";
  } else if (input === "info") {
    return "note";
  } else {
    // NOTE: In practice, this should never happen.
    return "none";
  }
}

/**
 * Convert a severity to a mock score; this value is used
 * by GitHub to calculate its own severity type
 */
function severityToScore(input: Severity): number {
  switch (input) {
    case "critical":
      return 9.5;
    case "high":
      return 8.0;
    case "medium":
      return 5.5;
    case "low":
      return 2.0;
    case "info":
      return 0;
  }
}

/**
 *  Convert a Platform report to Sarif.
 *
 * The data **must** be provided from the query in @mobsec/platform-lib with includeReport true,
 * otherwise data may be missing, in which case the behavior of this function
 * is undefined.
 *
 * @param assessment The assessment data from Platform
 * @param filter Filter parameters controlling which findings are reported
 * @param codePath Path to the source code; may be null
 * @param keyParams Controls generation of rule IDs
 * @param labUrl Platform UI server
 * @param errorLog Callback for error logging
 * @returns A SARIF report
 */
export async function convertToSarif(
  assessment: Assessment,
  filter: Filter = DEFAULT_FILTER,
  codePath = "./",
  keyParams: KeyParams = DEFAULT_KEY_PARAMS,
  labUrl: string = DEFAULT_LAB_UI_URL,
  errorLog?: (text: string) => void
): Promise<Log> {
  const report = assessment.report;

  if (!report) {
    throw new Error("No report data");
  }

  report.findings = report.findings.filter((finding) =>
    findingMatchesFilter(finding, filter)
  );

  const rules: ReportingDescriptor[] = [];
  for (const finding of report.findings) {
    if (!finding.affected) continue;

    let markdown = "";
    let issueSummary = "No issue description available.\n";
    if (finding.summary) {
      issueSummary = finding.summary;
    }
    const tags: string[] = [];

    const issue = finding.check.issue;
    if (issue) {
      if (issue.category) {
        tags.push(issue.category);
      }

      if (issue.cvss) {
        tags.push(`CVSS-${issue.cvss.toFixed(2)}`);
      }

      if (issue.impactSummary) {
        markdown += `## Business Impact\n${issue.impactSummary}\n`;
      }

      if (issue.stepsToReproduce) {
        markdown += `## Steps to Reproduce\n${issue.stepsToReproduce}\n`;
      }

      if (issue.recommendation || issue.codeSamples || issue.guidanceLinks) {
        markdown += "## Remediation Resources\n";
      }

      if (issue.recommendation) {
        markdown += `### Recommended Fix\n${issue.recommendation}\n`;
      }

      if (issue.codeSamples && issue.codeSamples.length !== 0) {
        markdown += "## Code Samples\n";

        for (const codeSample of issue.codeSamples) {
          markdown += "<details>\n";
          markdown += `<summary>${codeSample.caption} (click to expand)</summary>\n\n`;
          markdown += "```";
          if (codeSample.syntax) {
            markdown += codeSample.syntax;
          }
          markdown += "\n";
          markdown += codeSample.block;
          markdown += "\n```\n";
          markdown += "</details>\n";
        }
      }

      markdown += "\n";

      if (issue.guidanceLinks && issue.guidanceLinks.length !== 0) {
        markdown += "## Additional Guidance\n";
        for (const guidanceLink of issue.guidanceLinks) {
          markdown += `* ${guidanceLink.caption} ${guidanceLink.url}\n`;
        }
      }

      markdown += "\n";
    }

    // Format MobSec evidence in a Markdown table, when possible.

    const context = finding.context;
    const meta = finding.check.context;
    if (context && context.items && context.items.length > 0) {
      // Markdown tables look like:
      // |A|B|C|
      let markdownHeader = "|";
      // |-|-|-|
      let markdownDivider = "|";
      // If the key for "A" is "a", then it will be the first
      // entry in the ordering list, so on...
      const ordering = [];

      for (const field of meta.fields) {
        const fieldTitle = field.title;
        markdownHeader += `${fieldTitle}|`;
        markdownDivider += `-|`;
        ordering.push(field.key);
      }

      markdown += markdownHeader + "\n";
      markdown += markdownDivider + "\n";

      // For each row, insert keys in the order specified in "ordering".
      for (const row of context.items) {
        markdown += "|";
        for (const key of ordering) {
          let data = row[key];
          if (typeof data !== "string") {
            data = JSON.stringify(data);
          }

          markdown += "```` " + data + " ````" + "|";
        }
        markdown += "\n";
      }
    }

    const { ruleId } = ruleIdAndFingerprint(assessment, finding, keyParams);
    rules.push({
      id: ruleId,
      name: finding.title,
      helpUri: assessmentLink(labUrl, assessment, finding),
      shortDescription: {
        text: finding.title,
      },
      fullDescription: {
        text: issueSummary,
      },
      properties: {
        problem: {
          severity: finding.severity,
        },
        tags,
        precision: "medium",
        // security-severity is a string (even though it's a number)
        "security-severity": severityToScore(finding.severity).toString(),
      },
      help: {
        // NOTE: In practice this should not display on the GitHub UI.
        text: "MobSec only provides recommendations in a Markdown format.",
        markdown,
      },
    });
  }

  const results: Result[] = [];
  for (const finding of report.findings) {
    if (!finding.affected) continue;

    const issue = finding.check.issue;

    let issueDescription = "No issue description available\n";
    if (issue && issue.description) {
      issueDescription = issue.description;
    }

    const level = severityToNotification(finding.severity);
    const { ruleId, partialFingerprints } = ruleIdAndFingerprint(
      assessment,
      finding,
      keyParams
    );

    // If we are missing a specialized result for a rule, show the "simple"
    // result that does not show detailed line number information (refer to the
    // evidence table instead).
    const simpleResult: Result = {
      ruleId,
      message: {
        // Markdown doesn't work here. We render our information in the "help"
        // field in the reporting descriptor.
        text: issueDescription,
      },
      level: level,
      locations: [
        {
          // We don't have line number information so instead produce phony
          // information for a file that does not exist.
          physicalLocation: {
            artifactLocation: {
              uri: "Unknown",
              uriBaseId: "%SRCROOT%",
            },
            region: {
              startLine: 1,
              endLine: 1,
              byteOffset: 0,
              byteLength: 0,
            },
          },
        },
      ],
    };

    if (partialFingerprints) {
      simpleResult.partialFingerprints = partialFingerprints;
    }

    // FIXME: This should be refactored out and abstracted.
    const localResults = [];

    if (finding.key == "path_traversal" && codePath) {
      const context = finding.context;
      if (context && context.items && context.items.length > 0) {
        for (const row of context.items) {
          const name = row.name as string;
          const namespace = name.split(".");
          if (namespace.length > 0) {
            const [className] = namespace.slice(-1);
            // NOTE: Source mapping works for Java and Kotlin ONLY.
            const locations = await searchLocations(
              codePath,
              `class ${className}`,
              errorLog
            );

            for (const physicalLocation of locations) {
              const localResult: Result = {
                ruleId: ruleId,
                message: {
                  text: issueDescription,
                },
                level: level,
                locations: [
                  {
                    physicalLocation,
                  },
                ],
              };
              if (partialFingerprints) {
                localResult.partialFingerprints = partialFingerprints;
              }
              localResults.push(localResult);
            }
          }
        }
      }
    }

    if (localResults.length !== 0) {
      results.push(...localResults);
    } else {
      results.push(simpleResult);
    }
  }

  const run: Run = {
    tool: {
      driver: {
        name: "MobSec",
        informationUri: "https://www.khulnasoft.com/",
        semanticVersion: "1.0.0",
        rules,
      },
    },
    results,
  };

  const log: Log = {
    $schema: SARIF_SCHEMA_URL,
    version: "2.1.0",
    runs: [run],
  };

  return log;
}

function strip(name: string): string {
  return name.replace(/[^0-9a-z-_\s]/gi, "");
}

/*
 * Search the codebase for a string and return a physical location that corresponds
 * to it.
 */
async function searchLocations(
  codePath: string,
  name: string,
  errorLog?: (text: string) => void
): Promise<PhysicalLocation[]> {
  const errLog = errorLog || console.error;
  let results;
  try {
    results = await rg(codePath, `"${strip(name)}"`);
  } catch (e) {
    const error = e as RipGrepError;
    errLog(`Error: ripgrep: ${error.message}: ${error.stderr}`);
    errLog(
      "Note: ripgrep is required for line-number identification. On Ubuntu-based distributions use `apt-get install -y ripgrep` before running the MobSec action"
    );
    return [];
  }

  const locations = [];

  for (const result of results) {
    locations.push({
      artifactLocation: {
        uri: result.path.text,
        uriBaseId: "%SRCROOT%",
      },
      region: {
        startLine: result.line_number,
        endLine: result.line_number,
        byteOffset: 0,
        byteLength: 0,
      },
    });
  }

  return locations;
}
