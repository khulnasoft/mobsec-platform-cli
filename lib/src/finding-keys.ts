/*
 * Copyright Â© 2023 KhulnaSoft Ltd.
 *
 * SPDX-License-Identifier: MIT
 */

import crypto from "crypto";

import { Assessment, Finding } from "./types";
import { ValueError } from "./errors";
import { JSONObject } from "./client";

export interface KeyParams {
  fingerprint?: boolean;
  includePackage?: boolean;
  includePlatform?: boolean;
  v1platform?: "ios" | "android" | null;
  v1package?: string | null;
}

export const DEFAULT_KEY_PARAMS: KeyParams = {
  fingerprint: true,
  includePackage: true,
  includePlatform: true,
  v1package: null,
  v1platform: null,
};

/**
 * Take the SHA256 of an input string and output in hex.
 */
export function sha256(input: string): string {
  return crypto.createHash("sha256").update(input).digest("hex");
}

/**
 * Derive a key for a finding
 */
export function findingKey(
  assessment: Assessment,
  finding: Finding,
  keyParams: KeyParams
) {
  if (
    assessment.packageKey === keyParams.v1package &&
    assessment.platformType === keyParams.v1platform
  ) {
    return sha256(finding.key);
  }

  const keyInput = [finding.key];
  if (keyParams.includePlatform) {
    keyInput.push(assessment.platformType);
  }
  if (keyParams.includePackage) {
    keyInput.push(assessment.packageKey);
  }
  return sha256(keyInput.join("|"));
}

export type RuleIdAndFingerprint = {
  ruleId: string;
  partialFingerprints?: Record<string, string>;
};

export function ruleIdAndFingerprint(
  assessment: Assessment,
  finding: Finding,
  keyParams: KeyParams
): RuleIdAndFingerprint {
  if (!keyParams.fingerprint) {
    return {
      ruleId: findingKey(assessment, finding, keyParams),
    };
  }

  if (!(keyParams.includePlatform || keyParams.includePackage)) {
    return {
      ruleId: sha256(finding.key),
    };
  }

  const keyInput = [];
  if (keyParams.includePlatform) {
    keyInput.push(assessment.platformType);
  }
  if (keyParams.includePackage) {
    keyInput.push(assessment.packageKey);
  }
  return {
    ruleId: sha256(finding.key),
    partialFingerprints: {
      "appId/v1": sha256(keyInput.join("|")),
    },
  };
}

export function parseKeyConfig(keyData: JSONObject): KeyParams {
  const keyParams: KeyParams = { ...DEFAULT_KEY_PARAMS };

  const v1Key = keyData["v1-key"];
  if (v1Key !== undefined) {
    if (typeof v1Key !== "string") {
      throw new TypeError("v1-key must be a string");
    }
    const parts = v1Key.split(/\s+/);
    if (parts.length !== 2) {
      throw new ValueError(
        "v1-key must be of the form <platform> <packageName>"
      );
    }
    if (parts[0] !== "android" && parts[0] !== "ios") {
      throw new ValueError(
        `v1-key: "${parts[0]}" is not a valid platform type`
      );
    }
    [keyParams.v1platform, keyParams.v1package] = parts;
  }

  if ("package" in keyData) {
    if (typeof keyData.package !== "boolean") {
      throw new TypeError("package must be a boolean");
    }
    keyParams.includePackage = keyData.package;
  }

  if ("platform" in keyData) {
    if (typeof keyData.platform !== "boolean") {
      throw new TypeError("platform must be a boolean");
    }
    keyParams.includePlatform = keyData.platform;
  }

  if ("fingerprint" in keyData) {
    if (typeof keyData.fingerprint !== "boolean") {
      throw new TypeError("fingerprint must be a boolean");
    }
    keyParams.fingerprint = keyData.fingerprint;
  }

  return keyParams;
}
