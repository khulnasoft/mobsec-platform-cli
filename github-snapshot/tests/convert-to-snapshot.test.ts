/*
 * Copyright Â© 2022 MobSec Inc.
 *
 * SPDX-License-Identifier: MIT
 */

import { convertToSnapshot } from "../src/index";
import type { DependenciesOutput } from "@mobsec/platform-lib/lib/types";
import data from "./resources/dependencies-output.json";

describe("Snapshot conversion", () => {
  test("can perform conversion", async () => {
    const snapshot = convertToSnapshot(
      data as DependenciesOutput,
      "my-correlation-id",
      "7fc5849154e48c752bd5436f8ec32b2bc00e548e",
      "branch-or-tag",
      12345
    );
    expect(snapshot).toMatchSnapshot();
  });
});
