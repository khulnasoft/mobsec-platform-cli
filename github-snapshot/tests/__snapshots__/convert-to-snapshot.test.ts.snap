// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot conversion can perform conversion 1`] = `
Object {
  "detector": Object {
    "name": "MobSec",
    "url": "https://www.khulnasoft.com/",
    "version": "2.4.0",
  },
  "job": Object {
    "correlator": "my-correlation-id",
    "id": "12345",
  },
  "manifests": Object {
    "Payload/Example.app/ExampleFramework/Info.plist": Object {
      "name": "Payload/Example.app/ExampleFramework/Info.plist",
      "resolved": Object {
        "ExamplePod3": Object {
          "package_url": "pkg:cocoapods/ExamplePod3@3.0.0",
          "relationship": "direct",
        },
      },
    },
    "Payload/Example.app/libexample.dylib": Object {
      "name": "Payload/Example.app/libexample.dylib",
      "resolved": Object {
        "libexample": Object {
          "package_url": "pkg:generic/libexample@1.0.0",
          "relationship": "direct",
        },
      },
    },
    "Podfile.lock": Object {
      "name": "Podfile.lock",
      "resolved": Object {
        "ExamplePod1": Object {
          "package_url": "pkg:cocoapods/ExamplePod1@1.0.0",
          "relationship": "direct",
        },
        "ExamplePod2": Object {
          "package_url": "pkg:cocoapods/ExamplePod2@2.0.0",
          "relationship": "direct",
        },
      },
    },
  },
  "ref": "branch-or-tag",
  "scanned": "2022-01-01T00:00:00.000Z",
  "sha": "7fc5849154e48c752bd5436f8ec32b2bc00e548e",
  "version": 0,
}
`;
